### PR TITLE
palmerlr code review 2

### PR DESCRIPTION
**Pull request by liampalmer0**
# Palmerlr Code Review 2
## Analysis
The program was available to me on time and it compiles/runs fine.

The program has a bit more functionality than it did with my last review, specifically on the BMI page. The intent of the page is easy to see, it calculates a user's BMI based on their input. The code doesn't have many comments but the descriptive variable and function naming was sufficient for me to understand everything.

For my changes in this code review I focused on the code for the BMI Activity. The first thing I noticed was the large number of findViewById calls, so I thought it would be a good candidate for using view binding.

### Add Jetpack View Binding
Added view binding library which can replace calls to findViewById. It is type safe and null safe.
By adding this feature I was able to replace the large number of global variable declarations and findViewById calls in `BMIActivity.kt`.

### Refactor BMI Activity Logic
There was some code duplication and a few single use variables in `BMIActivity` so I refactored it to be more succinct.

### Rename confusing views
In the layout file for the BMI Activity, I noticed that many of the view ids were prefixed with 'activity_main_' which is confusing since they're not in the main activity layout. I decided to remove the prefixes altogether.

## Technical Concepts
1. Companion Objects
    - In some of the BMI related code I saw this group using companion objects. Since I am still new to Kotlin, this gave me the opportunity to see how a companion object could be used/implemented.
    - Example:
    ```
    companion object {
        val instance = BMICalcUtil()
        private const val CENTIMETERS_IN_METER = 100
        ...
        const val BMI_CATEGORY_UNDERWEIGHT = "Underweight"
        ...
    }
    ```
    
2. Toasts
    - I didn't have any previous experience with toasts but now I've seen examples of how simple they are to implement.
    - Example:
    ```
    Toast.makeText(this@BMIActivity, "Populate Weight and Height to Calculate BMI", Toast.LENGTH_SHORT).show()
    ```
3. Kotlin `when`
    - I noticed this type of conditional used in some of the code and didn't know what it was. So I researched it and now I know that it's used like a switch statement
    - Example:
    ```
    when (bmiCategory) {
        BMI_CATEGORY_UNDERWEIGHT -> bmiResultCardView.setCardBackgroundColor(Color.YELLOW)
        BMI_CATEGORY_HEALTHY -> bmiResultCardView.setCardBackgroundColor(Color.BLACK)
        ...
    }
    ```

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -30,6 +30,9 @@ android {
     kotlinOptions {
         jvmTarget = '1.8'
     }
+    buildFeatures {
+        viewBinding true
+    }
```
Comment (liampalmer0) [inline app/build.gradle:35]: Add view binding library

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt [hunk 2]
```diff
@@ -1,101 +1,95 @@
...
-        weightLbsEditText = findViewById(R.id.activity_main_weightlbs)
-        heightFtEditText = findViewById(R.id.activity_main_heightfeet)
-        heightInEditText = findViewById(R.id.activity_main_heightinches)
-        calculateButton = findViewById(R.id.activity_main_calculate)
-        toggleUnitsButton = findViewById(R.id.activity_main_toggleunits)
-        bmiTextView = findViewById(R.id.activity_main_bmi)
-        categoryTextView = findViewById(R.id.activity_main_category)
-        bmiResultCardView = findViewById(R.id.activity_main_resultcard)
+        binding = ActivityBmiCalculatorBinding.inflate(layoutInflater)
+        setContentView(binding.root)
```
Comment (liampalmer0) [inline app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt:23]: General implementation for view binding in an activity

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt [hunk 3]
```diff
@@ -1,101 +1,95 @@
...
     }
 
+    private fun showValidationMessage() {
+        Toast.makeText(
+            this@BMIActivity,
+            "Populate Weight and Height to Calculate BMI",
+            Toast.LENGTH_SHORT
+        ).show()
+    }
+
```
Comment (liampalmer0) [inline app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt:82]: refactor identical toast creation code into separate function

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/MainActivity.kt [hunk 4]
```diff
@@ -3,14 +3,18 @@ package edu.uc.mille5bm.smartnutrition
...
         super.onCreate(savedInstanceState)
-        setContentView(R.layout.activity_main)
-        findViewById<View>(R.id.btn_nutrition).setOnClickListener { startActivity(Intent(applicationContext, NutritionInfoActivity::class.java)) }
-        findViewById<View>(R.id.btn_calorie).setOnClickListener { startActivity(Intent(applicationContext, CalorieCounterActivity::class.java)) }
-        findViewById<View>(R.id.btn_health).setOnClickListener { startActivity(Intent(applicationContext, BMIActivity::class.java)) }
+        binding = ActivityMainBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        binding.btnNutrition.setOnClickListener { startActivity(Intent(applicationContext, NutritionInfoActivity::class.java)) }
+        binding.btnCalorie.setOnClickListener { startActivity(Intent(applicationContext, CalorieCounterActivity::class.java)) }
+        binding.btnHealth.setOnClickListener { startActivity(Intent(applicationContext, BMIActivity::class.java)) }
```
Comment (liampalmer0) [inline app/src/main/java/edu/uc/mille5bm/smartnutrition/MainActivity.kt:18]: added binding in main activity too

File: app/src/main/res/layout/activity_bmi_calculator.xml [hunk 5]
```diff
@@ -14,7 +14,7 @@
         android:textSize="32sp"/>
 
     <ToggleButton
-        android:id="@+id/activity_main_toggleunits"
+        android:id="@+id/toggleunits"
```
Comment (liampalmer0) [inline app/src/main/res/layout/activity_bmi_calculator.xml:17]: removed all `activity_main_` prefixes to view ids in this file

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt [hunk 6]
```diff
@@ -1,101 +1,95 @@
...
-                    val heightInCms = heightCmEditText.getText().toString().toDouble()
-                    val weightInKgs = weightKgEditText.getText().toString().toDouble()
-                    val bmi = BMICalcUtil().calculateBMIMetric(heightInCms, weightInKgs)
-                    displayBMI(bmi)
+                    displayBMI(
+                        BMICalcUtil().calculateBMIMetric(
+                            binding.heightcm.text.toString().toDouble(),
+                            binding.weightkgs.text.toString().toDouble()
+                        )
+                    )
```
Comment (liampalmer0) [inline app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt:37]: refactor variables used once to just be inline

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt [hunk 7]
```diff
@@ -1,101 +1,95 @@
...
-        bmiResultCardView.setVisibility(View.GONE)
-        calculateButton.setOnClickListener(View.OnClickListener {
+        binding.resultcard.visibility = View.GONE
+        binding.calculate.setOnClickListener(View.OnClickListener {
             if (inMetricUnits) {
-                if (weightKgEditText.length() == 0 || heightCmEditText.length() == 0) {
-                    Toast.makeText(this@BMIActivity, "Populate Weight and Height to Calculate BMI", Toast.LENGTH_SHORT).show()
+                if (binding.weightkgs.length() == 0 || binding.heightcm.length() == 0) {
+                    showValidationMessage()
                 } else {
```
Comment (liampalmer0) [inline app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt:31]: refactor duplicated toast creation code into function `showValidationMessage()`

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt [hunk 8]
```diff
@@ -1,101 +1,95 @@
...
+                            binding.weightkgs.text.toString().toDouble()
+                        )
+                    )
                 }
             } else {
-                if (weightLbsEditText.length() == 0 || heightFtEditText.length() == 0 || heightInEditText.length() == 0) {
-                    Toast.makeText(this@BMIActivity, "Populate Weight and Height to Calculate BMI", Toast.LENGTH_SHORT).show()
+                if (binding.weightlbs.length() == 0 || binding.heightfeet.length() == 0 || binding.heightinches.length() == 0) {
+                    showValidationMessage()
                 } else {
```
Comment (liampalmer0) [inline app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt:42]: refactor duplicated toast creation code into function `showValidationMessage()`

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt [hunk 9]
```diff
@@ -1,101 +1,95 @@
-
 package edu.uc.mille5bm.smartnutrition
 
 import android.graphics.Color
 import android.os.Bundle
-import androidx.appcompat.app.AppCompatActivity
 import android.view.View
-import android.widget.*
-import androidx.cardview.widget.CardView
+import android.widget.Toast
```
Comment (discospiff) [inline app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt:6]: explicit imports are more clear than wildcard.  Good idea.

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt (hunk 2 above)
Comment (discospiff) [reply app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt:23]: Ah, neat.  This saves a lot of code.

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt [hunk 10]
```diff
@@ -1,101 +1,95 @@
...
-            heightCmEditText.visibility = View.VISIBLE
-            weightKgEditText.visibility = View.VISIBLE
-            heightFtEditText.visibility = View.GONE
-            heightInEditText.visibility = View.GONE
-            weightLbsEditText.visibility = View.GONE
+            binding.heightcm.visibility = View.VISIBLE
+            binding.weightkgs.visibility = View.VISIBLE
+            binding.heightfeet.visibility = View.GONE
+            binding.heightinches.visibility = View.GONE
+            binding.weightlbs.visibility = View.GONE
```
Comment (discospiff) [inline app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt:65]: This does look a bit more readable.

File: app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt (hunk 3 above)
Comment (discospiff) [reply app/src/main/java/edu/uc/mille5bm/smartnutrition/BMIActivity.kt:82]: Good idea.  DRY principle.

File: app/src/main/res/layout/activity_bmi_calculator.xml (hunk 5 above)
Comment (discospiff) [reply app/src/main/res/layout/activity_bmi_calculator.xml:17]: That makes sense.  That is given by the context.

**Comment by discospiff** [comment]
Several good suggestions that reduce technical debt.  This is a good merge candidate.

**Pull request merged**